src/codeCache.cpp

```diff
@@ -311,3 +306,57 @@ size_t CodeCache::usedMemory() {
     }
     return bytes;
 }
+
+bool CodeCache::isValidHandle(void* handle) const {
+#ifdef __linux__
+    struct link_map* map;
+    // validate that the current loaded library is the same library that was observed during the /proc/self/maps processing
+    if (handle != NULL && dlinfo(handle, RTLD_DI_LINKMAP, &map) == 0) {
+        return _image_base == (const char*)map->l_addr;
+    }
+    return false;
+#else
+    return handle != NULL;
+#endif
+}
+
+UnloadProtection::~UnloadProtection() {
+    if (_lib_handle) {
+        dlclose(_lib_handle);
+    }
+}
+
+UnloadProtection::UnloadProtection(CodeCache *cc) {
+    _protected_cc = nullptr;
+    _lib_handle = nullptr;
+
+    if (Symbols::isMainExecutable(cc->_image_base, cc->_max_address) || Symbols::isLoader(cc->_image_base)) {
+        _protected_cc = cc;
+        return;
+    }
+
+    // Protect library from unloading while parsing in-memory ELF program headers.
+    // Also, dlopen() ensures the library is fully loaded.
+    void* handle_ptr = dlopen(cc->name(), RTLD_LAZY | RTLD_NOLOAD);
+    if (cc->isValidHandle(handle_ptr)) {
```

Comment:
Should we log a warning if dlopen fails here, except for the main executable and the loader?

Reply:
I think we should rather log one time in `patchImport`